File: backend/services/stock_service.py
import yfinance as yf
from typing import Dict, Optional
import logging

logger = logging.getLogger(__name__)


class StockService:
    """주가 정보 조회 서비스 (yfinance)"""
    
    @staticmethod
    def get_current_price(ticker: str) -> Optional[float]:
        """현재가 조회"""
        try:
            stock = yf.Ticker(ticker)
            info = stock.info
            
            # 현재가 또는 최근 종가
            current_price = info.get('currentPrice') or info.get('regularMarketPrice')
            return float(current_price) if current_price else None
        except Exception as e:
            logger.error(f"Error fetching price for {ticker}: {e}")
            return None
    
    @staticmethod
    def get_stock_info(ticker: str) -> Dict:
        """종목 상세 정보 조회"""
        try:
            stock = yf.Ticker(ticker)
            info = stock.info
            
            return {
                "ticker": ticker,
                "name": info.get('longName') or info.get('shortName', ticker),
                "current_price": info.get('currentPrice') or info.get('regularMarketPrice'),
                "previous_close": info.get('previousClose'),
                "daily_change": info.get('regularMarketChangePercent'),
                "currency": info.get('currency', 'USD')
            }
        except Exception as e:
            logger.error(f"Error fetching info for {ticker}: {e}")
            return {
                "ticker": ticker,
                "name": ticker,
                "current_price": None,
                "previous_close": None,
                "daily_change": None,
                "currency": "USD"
            }
    
    @staticmethod
    def get_previous_close(ticker: str) -> Optional[float]:
        """전일 종가 조회"""
        try:
            stock = yf.Ticker(ticker)
            return float(stock.info.get('previousClose', 0))
        except Exception as e:
            logger.error(f"Error fetching previous close for {ticker}: {e}")
            return None
    
    @staticmethod
    def calculate_change_percent(current: float, previous: float) -> float:
        """가격 변동률 계산"""
        if previous == 0:
            return 0.0
        return ((current - previous) / previous) * 100

File: backend/src/common/services/stock.service.ts
import { Injectable, Logger } from '@nestjs/common';
import axios from 'axios';

@Injectable()
export class StockService {
  private readonly logger = new Logger(StockService.name);

  async getCurrentPrice(ticker: string): Promise<number | null> {
    try {
      // yfinance 대신 Yahoo Finance API 사용
      const url = `https://query1.finance.yahoo.com/v8/finance/chart/${ticker}`;
      const response = await axios.get(url);
      
      const quote = response.data.chart.result[0].meta;
      return quote.regularMarketPrice || null;
    } catch (error) {
      this.logger.error(`Failed to fetch price for ${ticker}:`, error.message);
      return null;
    }
  }

  async getStockInfo(ticker: string) {
    try {
      const url = `https://query1.finance.yahoo.com/v8/finance/chart/${ticker}`;
      const response = await axios.get(url);
      
      const result = response.data.chart.result[0];
      const quote = result.meta;
      
      return {
        ticker,
        name: quote.longName || quote.shortName || ticker,
        currentPrice: quote.regularMarketPrice,
        previousClose: quote.previousClose,
        dailyChange: quote.regularMarketChangePercent,
      };
    } catch (error) {
      this.logger.error(`Failed to fetch info for ${ticker}:`, error.message);
      return {
        ticker,
        name: ticker,
        currentPrice: null,
        previousClose: null,
        dailyChange: null,
      };
    }
  }

  async getPreviousClose(ticker: string): Promise<number | null> {
    try {
      const url = `https://query1.finance.yahoo.com/v8/finance/chart/${ticker}`;
      const response = await axios.get(url);
      
      const quote = response.data.chart.result[0].meta;
      return quote.previousClose || null;
    } catch (error) {
      this.logger.error(`Failed to fetch previous close for ${ticker}:`, error.message);
      return null;
    }
  }

  calculateChangePercent(current: number, previous: number): number {
    if (previous === 0) return 0;
    return ((current - previous) / previous) * 100;
  }
}